src/field_theory/ratpoly.lean

```diff
@@ -0,0 +1,391 @@
+/-
+Copyright (c) 2021 Anne Baanen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Anne Baanen
+-/
+
+import ring_theory.localization
+
+/-!
+# The field of rational polynomials
+
+This file defines the field `ratpoly K` of rational polynomials over a field `K`,
+and shows it is the field of fractions of `polynomial K`.
+
+## Implementation notes
+
+We need a couple of maps to set up the `field` and `is_fraction_ring` structure,
+namely `ratpoly.of_fraction_ring`, `ratpoly.to_fraction_ring`, `ratpoly.mk` and `ratpoly.aux_equiv`.
+All these maps get `simp`ed to bundled morphisms like `algebra_map (polynomial K) (ratpoly K)`
+and `is_localization.alg_equiv`.
+-/
+
+noncomputable theory
+open_locale classical
+open_locale non_zero_divisors
+
+universes u v
+
+variables (K : Type u) [field K]
+
+/-- `ratpoly K` is `K(x)`, the field of rational polynomials over `K`.
+
+The inclusion of polynomials into `ratpoly` is `algebra_map (polynomial K) (ratpoly K)`,
+the maps between `ratpoly K` and another field of fractions of `polynomial K`,
+especially `fraction_ring (polynomial K)`, are given by `is_localization.algebra_equiv`.
+-/
+structure ratpoly : Type u := of_fraction_ring ::
```

Comment:
Wouldn't it be cheaper (in terms of lines of code) to make an irreducible def that deduces all the algebraic structure using some derive handler?

Reply:
The drawback is that `derive` produces a (semi)reducible instances, and we want to have irreducible ones if we want the implementation in terms of `fraction_ring` to be hidden. More precisely, we want fields carrying data inside the instances to be irreducible defs; irreducible instances is apparently considered harmful.

Reply:
Hmm, I see. Sad thing that we need so much "trivial" code.